### PR TITLE
SCIM: User get all filters

### PR DESCRIPTION
**Pull request by vdavid**
⚠️ **Tip for reviewers**: skip the `/filter` folder! Only check our two files!

**Whoa**, now, this was some hardcore coding.

This PR adds **complete** filter support for User `GET`.

By “complete”, I mean it handles _insane_ queries like
`filter=emails[type eq "work" and value co "@example.com"] or ims[type eq "xmpp" and value co "@foo.com"]`.
It's meant to be 100% compatible with the RFC: https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.2, plus it's thoroughly tested.

Did I code it all by myself? Nope. I found the logic in the Elimity Go package ([here](https://sourcegraph.com/github.com/elimity-com/scim@15165b1a61c853d4e5e9ccfa54cfb2efc92eb733/-/blob/internal/filter/filter.go?L120) – found it using Sourcegraph search!), which is btw the same package that we use for our endpoint. But the filtering code is in the `/internal` folder of the package, and the only public part that uses it is the `/Schemas` REST endpoint handler and their test.
So, as it's very well-written logic, plus it's under the MIT license, I copy&pasted it to our code, and I use it for the filtering.
This should be legally fine, and as for elegance, I found no better way.

**Caveat:** as the filtering is done with **Go code**, _all_ resources (means: _all_ users) are fetched from the DB, then filtered in the backend. This is a limitation, as it will use more than a reasonable about of memory and processing power for customers with hundreds of thousands of users. Thoughts about this:
 - I've added a shortcut of “If there's no filter, just use and SQL with an OFFSET and LIMIT”. So the handler will only take the slow path if there's a filter.
 - We can add more shortcuts like this, like “If there is only a simple `userName` filter, just use SQL”. This is what the devs of another project did [here](https://sourcegraph.com/github.com/finleap-connect/monoskope@66c89aa22ad6903973591d2b58891e4872aa0c59/-/blob/internal/scimserver/user_handler.go?L162).
 - We can add a limitation that “under 1,000 users, we support complex filters; over that, we only support very basic ones”, and the limit could be configurable in the site config if customers really need this.
 - And then there's **the hard way**: we can rewrite the same logic with `sqlf` calls and transform the filters to an SQL query. I've tried to find an implementation for this but found none with Sourcegraph. I don't think it's needed at this point, though.
 - Btw, filter support is an _optional_ feature in SCIM (like most things in SCIM, tbh), so this is not a hard requirement. But it's quite close to the core requirements, and the current solution was worth a half day of fun engineering work, I'd say.

Another goodie for the end: this implementation is not specific to the resource type, so we'll get filtering for Groups for free once we add Group support later.

## Test plan

The new code is well tested (both the library I imported and my own part); I'm confident about it. Plus, customers won't use it yet.

### CODE REVIEW

File: enterprise/internal/scim/user.go [hunk 1]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
...
 // Page.Resources. Otherwise, is an empty slice is assigned, an empty list will be represented as `[]`.
 func (h *UserResourceHandler) GetAll(r *http.Request, params scim.ListRequestParams) (scim.Page, error) {
+	var totalCount int
+	var resources []scim.Resource
+	var err error
+
+	if params.Filter == nil {
+		totalCount, resources, err = h.getAllFromDB(r, params.StartIndex, &params.Count)
+	} else {
+		// Fetch all resources from the GB and then filter them here.
```
Comment (pjlast) [inline enterprise/internal/scim/user.go:86]:
```suggestion
		// Fetch all resources from the DB and then filter them here.
```

I assume this is supposed to read DB?

File: enterprise/internal/scim/user.go [hunk 2]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
 // An empty list of resources will be represented as `null` in the JSON response if `nil` is assigned to the
 // Page.Resources. Otherwise, is an empty slice is assigned, an empty list will be represented as `[]`.
```
Comment (pjlast) [inline enterprise/internal/scim/user.go:77]:
```suggestion
// Page.Resources. Otherwise, if an empty slice is assigned, an empty list will be represented as `[]`.
```

Not really part of the PR but saw it :P

File: enterprise/internal/scim/user.go [hunk 3]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
...
+}
+
+func (h *UserResourceHandler) getAllFromDB(r *http.Request, startIndex int, count *int) (totalCount int, resources []scim.Resource, err error) {
 	// Get total count
-	totalCount, err := h.db.Users().Count(r.Context(), &database.UsersListOptions{})
+	totalCount, err = h.db.Users().Count(r.Context(), &database.UsersListOptions{})
 	if err != nil {
-		return scim.Page{}, err
+		return
 	}
```
Comment (pjlast) [inline enterprise/internal/scim/user.go:123]: I think we can have a slight optimisation here and only do a query for `totalCount` like this if `startIndex` and `count` is set? Because otherwise it's going to fetch all of the users anyways and you can just return `len(users)` as the totalCount.

File: enterprise/internal/scim/user.go [hunk 4]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
...
-			Offset: offset,
-		},
-	})
+	var opt = &database.UsersListOptions{}
+	if count != nil {
+		opt = &database.UsersListOptions{
+			LimitOffset: &database.LimitOffset{Limit: *count, Offset: offset},
+		}
+	}
+	users, err := h.db.Users().ListForSCIM(r.Context(), opt)
```
Comment (pjlast) [inline enterprise/internal/scim/user.go:138]: Just confirming it's intentional to not allow `offset` without `count`? In which case I would move the `offset` calculation inside this if statement.

**Review by pjlast — APPROVED** [review]
Just some nits :)

And I would maybe just ask for an extra sentence or two in `filter.go` explaining why we're copying the code from the package instead of importing the package and using it?

File: enterprise/internal/scim/user.go [hunk 5]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
...
+		// Fetch all resources from the GB and then filter them here.
+		// This doesn't feel efficient, but it wasn't reasonable to implement this in SQL in the time available.
+		var allResources []scim.Resource
+		_, allResources, err = h.getAllFromDB(r, 0, nil)
+
+		for _, resource := range allResources {
+			extensionSchemas := make([]schema.Schema, 0, len(h.schemaExtensions))
+			for _, ext := range h.schemaExtensions {
+				extensionSchemas = append(extensionSchemas, ext.Schema)
+			}
```
Comment (mrnugget) [inline enterprise/internal/scim/user.go:95]: Do we need to do this in every iteration of the loop? `h.schemaExtensions` is static, right? So we could do this only once and then reuse it.

File: enterprise/internal/scim/user.go [hunk 6]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
...
+
+			validator := filter.NewFilterValidator(params.Filter, h.coreSchema, extensionSchemas...)
```
Comment (mrnugget) [inline enterprise/internal/scim/user.go:97]: Same goes for `validator`: does it keep state? If not, can we initialize it once and only call it inside loop iteration?

File: enterprise/internal/scim/user.go [hunk 7]
```diff
@@ -70,38 +76,75 @@ func (h *UserResourceHandler) Get(r *http.Request, idStr string) (scim.Resource,
...
+			if err := validator.PassesFilter(resource.Attributes); err != nil {
+				continue
+			}
+
+			totalCount++
+			if totalCount >= params.StartIndex && len(resources) < params.Count {
```
Comment (mrnugget) [inline enterprise/internal/scim/user.go:103]: Can't you abort the loop once `len(resources) == params.Count`? Otherwise you keep iterating over all users even though `len(resources)` is already full, right?

**Review by mrnugget — APPROVED** [review]

File: enterprise/internal/scim/user.go (hunk 7 above)
Comment (pjlast) [reply enterprise/internal/scim/user.go:103]: So the loop needs to continue so that `totalCount` equals the total number of filtered users

File: enterprise/internal/scim/user.go (hunk 3 above)
Comment (vdavid) [reply enterprise/internal/scim/user.go:123]: That'd make sense, but `count` is a required field. It defaults to `100`.
It might be `<= 0`, in which case we return 0 results, but even in that case, we need the `totalCount`.

File: enterprise/internal/scim/user.go (hunk 4 above)
Comment (vdavid) [reply enterprise/internal/scim/user.go:138]: It's intentional.
`count == nil` is only used in one, artificial case:
When we get a filtering expression (`filter != nil`) then we want to get _all_ DB records. That's the only case when `count == nil`, and then we don't want to do any offsetting because when we fetch the data we have no idea how many records the filters will let through.

File: enterprise/internal/scim/user.go (hunk 5 above)
Comment (vdavid) [reply enterprise/internal/scim/user.go:95]: Yup, thanks! I spotted this at one point, but then I got sidetracked with something else and never did this change.

File: enterprise/internal/scim/user.go (hunk 7 above)
Comment (vdavid) [reply enterprise/internal/scim/user.go:103]: As @pjlast said. This is actually a mistake that the test server in the `elimity` package itself do (fyi, it's [here](https://sourcegraph.com/github.com/elimity-com/scim/-/blob/server_test.go?L156&subtree=true)), and then they end up returning the wrong results. (okay, it's not _exactly_ the same mistake, but it does return an incorrect number for `TotalResults`.)

Comment (mrnugget) [reply enterprise/internal/scim/user.go:103]: In that case I'd add a comment to explain that. It's something I can see someone else removing "by accident"

Comment (vdavid) [reply enterprise/internal/scim/user.go:103]: Well, the tests will fail as soon as someone removes it. So it has to be a very intentional change. That said, I'll add a comment in a follow-up PR to explain this
